rfcs/0081-dynmical-config-for-stream-job.md

```diff
@@ -0,0 +1,81 @@
+### Non-Goals
+
+The RFC will not concern those config can change the streaming graph and its topology, such as
+- The configs that can change the plan such as `RW_TWO_PHASE_AGG`
+- The configs that can change the actor's number such as `STREAMING_PARALLELISM` (maybe we can use the same syntax to alter it later, but it will be only a syntactic sugar of scale)
+
+Also, the RFC do not associate with those node-level config such as `exchange_initial_permits` 
+
+These are just some example configs that **must not** including in the framework, there still are some vague config need to be discussed case by case, which mainly depends on if it is needed to be different between different streaming jobs, such as `chunk_size`(If it is meaningful to have different chunk_size for each actors) and `connector_message_buffer_size`(It looks more like a configuration per source?).
+
+## Unified streaming job configs
+
+We will introduce "dynamic stream config" concept to the user. The config entries in that can be set in the with clause when creating and we have a lots of discussion in the issue https://github.com/risingwavelabs/risingwave/issues/11929.
+```SQL
+create materialized view mv with ( streaming_rate_limit = 200 );
+```
+
+detailed implementation, a `DynamicConfig` struct will be stored in the `TableFragments` just like current `StreamEnvironment`. And when creating the actors and executors inside, the executor will get their needed parameter for the config.
+Btw, I‘d like to still maintain the `StreamEnvironment` as the immutable parameters.
+
+### relation with session config
+
+Please notice that the config is independent with session config. In Conceptual space, we can define a dynamic stream config without its corresponding session variable.
+The session config of the corresponding dynamic stream config just define its default value, in other words, when the config entry is not defined in the with clause, what's the value should be.
+Furthermore, do we need add a copy of a session configs in the system parameters as the default value of the session config, that is another topic and will not talked in this RFC.
+
+## Alter dynamic configs
+
+First to clarify, there is some parameter can not be altered such as `timezone`. Because it can change the result of the streaming query in SQL semantic. It is strange to alter such a config for a "running query", it is a little like alter mv which change the SQL statement.
+
+User will be able to alter the dynamic stream config for specified MV or SINK with the grammar.
+```SQL
+ALTER [MATERIALIZED VIEW | SINK] SET parameter TO [value | 'value'];
+```
+The statement will send a RPC to meta. After reducing the config in the common place, it is easier to alter the config of some relations. And we need some implementation complexity trade-off here.
+
+1. alter with recovery: change the configs in the meta store and trigger a recovery. Then the newly created actor will adopt the new configs.
```

Comment:
changing config seems too heavy to apply. Though it is the easiest way, it is counter-intuitive that users change sth and get lost from the cluster, even for several seconds. 
I don't get the position for the proposal if it is an SQL command, I don't think it can alter the cluster's status.
If it is a DevOps tool, it should not be released as a command. I think meta command is a better way.

Reply:
>  get lost from the cluster, even for several seconds.
I don't think we need to trigger the recovery, we can let user manually trigger it.

The UX will be:
1. user alters the MV: `alter materialized view mv1 with ( ... )` (perhaps need a section on the exact syntax)
2. the user receives a response + a notice: `NOTICE: Please restart the cluster with TRIGGER RECOVERY / risectl meta recovery for your changes to take effect`
3. The user runs the recovery command.

So it would be explicit and opt-in by the user

Reply:
> I don't get the position for the proposal if it is an SQL command.

So how about the 4th method :thinking: do you think it is acceptable if we can change the log without recovery?

Reply:
yes, I think 4 is the ideal way for such change. I think there are two phases: 1) change the proto in etcd and  commit to meta, it is safe to restart the cluster, 2) tell the user that the change has been applied. 

> the user receives a response + a notice: NOTICE: Please restart the cluster with TRIGGER RECOVERY / risectl meta recovery for your changes to take effect

If so, we are encouraging the users to test with risectl, a not well-documented tool designed for advanced users and the dev team. I mean doing anything in SQL, then it should be simple and concise, without introducing extra tools or doing the devops' job.

Reply:
> > the user receives a response + a notice: NOTICE: Please restart the cluster with TRIGGER RECOVERY / risectl meta recovery for your changes to take effect
> 
> If so, we encourage users to test with risectl, a poorly well-documented tool designed for advanced users and the dev team. I mean doing anything in SQL, should be simple and concise, without introducing extra tools or doing the devops' job.

I guess it is more like the `ALTER SYSTEM` statement
>Values set with ALTER SYSTEM will be effective after the next server configuration reload, or after the next server restart in the case of parameters that can only be changed at server start. A server configuration reload can be commanded by calling the SQL function pg_reload_conf(), running pg_ctl reload, or sending a SIGHUP signal to the main server process.